wikitextprocessor/luaexec.py

```diff
@@ -574,22 +594,66 @@ def expandTemplate(frame, *args):
         # Ignore this error - it is an error but a clear error in Wiktionary
         # rather than in the extractor.
         return ""
-    elif "attempt to index a nil value (local 'lang')" in text and \
-         "in function 'Module:links.getLinkPage'" in text:
+    elif (
+        "attempt to index a nil value (local 'lang')" in text
+        and "in function 'Module:links.getLinkPage'" in text
+    ):
         # Ignore this error - happens when an unknown language code is passed
         # to various templates (a Wiktionary error, not extractor error)
         return ""
     else:
         if "check deprecated lang param usage" in ctx.expand_stack:
-            ctx.debug("LUA error but likely not bug -- in #invoke {} parent {}"
-                      .format(invoke_args, parent),
-                      trace=text, sortid="luaexec/679")
+            ctx.debug(
+                "LUA error but likely not bug -- in #invoke {} parent {}".format(
+                    invoke_args, parent
+                ),
+                trace=text,
+                sortid="luaexec/679",
+            )
         else:
-            ctx.error("LUA error in #invoke {} parent {}"
-                      .format(invoke_args, parent),
-                      trace=text, sortid="luaexec/683")
+            ctx.error(
+                "LUA error in #invoke {} parent {}".format(invoke_args, parent),
+                trace=text,
+                sortid="luaexec/683",
+            )
     msg = "Lua execution error"
     if "Lua timeout error" in text:
         msg = "Lua timeout error"
-    return ('<strong class="error">{} in Module:{} function {}'
-            '</strong>'.format(msg, html.escape(modname), html.escape(modfn)))
+    return (
+        '<strong class="error">{} in Module:{} function {}'
+        "</strong>".format(msg, html.escape(modname), html.escape(modfn))
+    )
+
+
+@functools.cache
+def query_wikidata(item_id: str):
+    import requests
+
+    r = requests.get(
```

Comment:
We don't query data in other places, so this is a little out of place. Ideally, we should just gather the data cached in this function, make it a separate dict or json file, and load the data from there, updating as needed.

We could do this by using an error/debug/warning message each time 'query_stored_wikidata' is called and it can't find the value in the previously stored data. The error messages can be extracted by a separate script that can then reimplement the functionality here to query data (using the old keys and error keys) and save it locally, until we get more "missing wikidata" errors. I also think all of this should probably happen in Wiktextract, it feels like most things having to do with data should go there and only passed on to wikitextprocessor.

Currently, this should work (afaict), so the above is for the future.

Reply:
The API results are cached(please see the above `@functools.cache` line).

Reply:
Yes, the results are cached when wikitextprocessor is running, but we probably do not want to query wikidata each time wikitextprocessor runs. We should pass data into wikitextprocessor from wiktextract (and most preferably local data without querying wikidata each time).